src/dragonball/docs/vcpu.md

```diff
@@ -0,0 +1,41 @@
+# vCPU
+
+## VCPU Manager
+The vcpu manager is to manage all vcpu related actions, we will dive into some of the important struct members in this doc.
+
+For now, aarch64 vCPU support is still under development, we'll introduce it when we merge `runtime-rs` to the master branch.
+
+### vCPU config
+`VcpuConfig` is used to configure guest overall cpu info.
+
+`boot_vcpu_count` is to define the initial vcpu number.
+
+`max_vcpu_count` is to define the maximum vcpu number and it's used for the upper boundary for cpu hotplug feature
+
+`thread_per_core`, `cores_per_die`, `dies_per_socket` and `socket` are for defining cpu topology.
+
+`vpmu_feature` is to define vpmu feature level.
+If `vpmu` feature is Disabled, it means `vpmu` feature is off (by default).
+If `vpmu` feature is LimitedlyEnabled, it means minimal vpmu counters are supported (cycles and instructions).
+If `vpmu` feature is FullyEnabled, it means all vpmu counters are supported
+
+## vCPU State
+
+There are four states for vcpu state machine: `running`, `paused`, `waiting_exit`, `exited`. There is a state machine to maintain the task flow.
+
+When the vCPU is created, it'll turn to `paused` state. After vCPU resource is ready at VMM, it'll send a `Resume` event to the vCPU thread, and then vCPU state will change to `running`.
+
+During the `running` state, vmm will try to catch vCPU exit and execute different logic according to the exit reason.
+
+If the VMM catch some exit reasons that it cannot handle, the state will change to `wait_exit` and VMM will try to stop the virtual machine.
+
+When the VMM is stopped / destroyed, the state will change to `exited`.
+   
+## vCPU Hot plug
+Since `Dragonball Sandbox` doesn't support virtualization of ACPI system, we use [upcall](https://github.com/openanolis/dragonball-sandbox/tree/main/crates/dbs-upcall) to establish a direct communication channel between `Dragonball` and Guest in order to trigger vcpu hotplug.
+
+To use `upcall`, kernel patches are needed, you can get the patches from [upcall](https://github.com/openanolis/dragonball-sandbox/tree/main/crates/dbs-upcall) page, and we'll provide a ready-to-use guest kernel binary for you to try.
```

Comment:
@jiangliu et al -- has there been discussion with Linux kernel community around these patches?

Reply:
Hi @egernst.) As @jiangliu mentioned in #4192, we are planning to open-source this driver, but we haven't started conversations with Linux kernel community yet.
And we will submit patches of this driver to kata into `kata-containers/tools/packaging/kernel/patches/` for users to use upcall.